content/en/contribute/code/core/automated-tests.md

```diff
@@ -115,3 +115,37 @@ Running e2e tests can be quite slow so to save time modify the `specs` property
 #### Watching the test run
 
 Running the tests locally with `npm run wdio-local` or `npm run standard-wdio-local` will allow you to watch it run but if you interact with the page the test will fail in unexpected ways. Furthermore the browser will close after a short timeout so you won't be able to inspect the console or DOM. To do this, force quit the process running the test before it tears down and you will be able to navigate around the app, use Chrome dev tools, and inspect the docs in the database to (hopefully) work out what's going wrong.
+
+#### Running the upgrade e2e test locally
+
+To run the upgrade e2e tests in your local environment, follow these steps:
+
+- Make sure your branch has been published, and it's available in the market:
+    - A way to do this is by pushing the branch, let the GitHubActions to run, if all the other e2e are okay, then it will publish the branch.
+    - Check that your branch name is available [here](https://staging.dev.medicmobile.org/_couch/builds_4/_design/builds/_view/releases).
+- Make sure to stop all existing containers
+- Set these environment variables:
+    - `export MARKET_URL_READ=https://staging.dev.medicmobile.org`.
+    - `export STAGING_SERVER=<builds path>`.
+      - To upgrade from CHT v4.x.x use `export STAGING_SERVER=_couch/builds_4`.
+      - To upgrade from previous versions of CHT use: `export STAGING_SERVER=_couch/builds`.
+    - `export BRANCH=<your branch name>`.
+- Start the containers for e2e tests like this
+  - TBD
```

Comment:
To start the container, for now, we need to add this: 
![Screenshot 2023-11-14 at 10 44 54 pm](https://github.com/medic/cht-docs/assets/66472237/1af262cf-ea63-4421-a98d-f9f7c314ecc1)

Reply:
@lorerod FYI

Reply:
@dianabarsan pinging you on this again. 
These changes allow the upgrade tests in the local environment to run, but they still fail while in the master's CI the pass. This probably isn't the correct way to run them. 

```
[0-0] AssertionError in "Performing an upgrade.should have an upgrade_log after installing"
AssertionError: expected +0 to equal 1
    at Context.<anonymous> (/Users/tlepiz/Documents/Medic/Repos/PRs/cht-core/tests/e2e/upgrade/upgrade.wdio-spec.js:82:28)
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
[0-0] Error in "Performing an upgrade.should upgrade to current branch"
Error: element ("div*=Deployment complete") still not displayed after 15000ms
    at async Context.<anonymous> (/Users/tlepiz/Documents/Medic/Repos/PRs/cht-core/tests/e2e/upgrade/upgrade.wdio-spec.js:103:5)
```
[Video from Tatiana](https://medic.slack.com/files/U02LRG2F8TG/F065TKML6LC/screen_recording_2023-11-14_at_2.38.05___pm.mov)

@tatilepizs @lorerod and I are thinking that we're missing some setup. The QA Team wants to learn and document how to run these tests in their local environment to debug whenever needed. 

What are we missing here? What is the best way to run them in the local environment?

Reply:
I went over the upgrade e2e test locally and everything worked. I made no changes to the upgrade e2e test, as suggested above, and those changes should not be necessary. 

My experience was:
- I needed to stop existing containers that cause port collisions 
- on slow internet, the tests failed because the images weren't downloaded in sufficient time, this just meant I had to restart the test a bunch of times until the images were downloaded
- I had to checkout and run the tests over the branch I wanted to upgrade to. This is mainly necessary to compute the correct version within the test. 

![image](https://github.com/medic/cht-docs/assets/35681649/c84363b3-f423-4626-a76a-3382a5e503a2)

Reply:
@dianabarsan can you please share the exact steps that you are doing to run the tests? 
I still need to run `npm run wdio-local` to have all the containers ready and then run `npm run upgrade-wdio` and it still fails 😞 

You pointed out that maybe on slow internet it will fail, how fast my internet needs to be, maybe my internet is too slow  😅. Is there a way to check if the images started the download process but they didn't finish? Because I ran the test 10 times and it is still failing. How many times do you think that it would require to be run to download all the images successfully?

[video](https://github.com/medic/cht-docs/assets/94494491/f068a99b-5b15-4ad7-8437-dd8b84a45ff0)

Reply:
Checkout a working branch `message-tab-link-to-contact` for testing purpose.
Check that the branch name is available [here](https://staging.dev.medicmobile.org/_couch/builds_4/_design/builds/_view/releases).
Define env variables:
 - `export MARKET_URL_READ=https://staging.dev.medicmobile.org`
 -  `export STAGING_SERVER=_couch/builds_4`
 - `export BRANCH=message-tab-link-to-contact`
 - Run `npm run upgrade-wdio`
 The only container that is created is `upgrade`. Non of the cht containers are created
 I saw what Diana mentioned:

```
Checking API
Error with request: GET https://localhost/api/info
API check failed, trying again in 1 second
Error: connect ECONNREFUSED 127.0.0.1:443
Services took too long to start. Shutting down...

      If you are seeing this locally, it can mean that your internet is too slow to download all images in the 
      allotted time. 
      Either run the test multiple times until you load all images, download images manually or increase this timeout.
    
Checking API
Error with request: GET https://localhost/api/info
API check failed, trying again in 1 second
Error: connect ECONNREFUSED 127.0.0.1:443
time="2023-11-21T18:49:33-03:00" level=warning msg="The \"COUCHDB_UUID\" variable is not set. Defaulting to a blank string."

time="2023-11-21T18:49:33-03:00" level=warning msg="The \"COUCHDB_UUID\" variable is not set. Defaulting to a blank string."
time="2023-11-21T18:49:33-03:00" level=warning msg="The \"COUCHDB_UUID\" variable is not set. Defaulting to a blank string."

 Volume cht-e2e_cht-ssl  Removing
 Volume cht-e2e_cht-credentials  Removing

 Volume cht-e2e_cht-ssl  Removed

 Volume cht-e2e_cht-credentials  Removed

docker-compose -f /var/folders/t9/pnvyjb694k7130mkv_126vyc0000gn/T/upgrade-service-OaLwDJ/cht-upgrade-service.yml -p upgrade down
```

I will try downloading images manually.

Reply:
- Manually downloaded the images using [cht-core.yml](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:message-tab-link-to-contact/docker-compose/cht-core.yml) and [cht-couchdb.yml](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:message-tab-link-to-contact/docker-compose/cht-couchdb.yml)
- Executing:

`COUCHDB_USER=admin COUCHDB_PASSWORD=pass docker-compose -f docker-compose.yml pull`
`COUCHDB_USER=admin COUCHDB_PASSWORD=pass docker-compose -f cht-couchdb.yml pull`

- I check if the images exist:

![image](https://github.com/medic/cht-docs/assets/21312057/7b86b717-60cb-4dc5-9f52-61de041b763f)

- Run `npm run upgrade-wdio`

The only container that is created is `upgrade`. Non of the cht containers are created
And got the same error:

```
Services took too long to start. Shutting down... 
If you are seeing this locally, it can mean that your internet is too slow to download all images in the allotted time. 
Either run the test multiple times until you load all images, download images manually or increase this timeout.
```

So definitely, I'm missing a step.
@tatilepizs Did you have any luck?
@dianabarsan Is there something that I might be overlooking?

Reply:
Can you please try also downloading the initial images for the latest release? The upgrade container initially tries to setup latest release of CHT and you could be missing those images as well. 

Also, please share the logs for the upgrade container. 
Thanks!

Reply:
Ok I think we are getting somewhere. 
Something to do with Python?
[upgrade.log](https://github.com/medic/cht-docs/files/13444487/upgrade.log)

Reply:
```
python --version 
Python 2.7.18
```

Reply:
This is strange. Do you otherwise use the upgrade service successfully?

Reply:
It is not working for me locally. It gives the same error as the video shared by Tatiana [here](https://github.com/medic/cht-docs/pull/1218/files#r1397693892)

Reply:
@lorerod Can you manually try using the upgrade service to upgrade from one version to another?

Reply:
For me, the upgrade service has only worked when I use the docker-helper script to have my instance up, if I use the development set-up it has never worked.

Reply:
Can you share your upgrade service logs please?

Reply:
I tried running the `npm run upgrade-wdio`first with my docker completley clean (no containers, images nor volumes) and the second time a downloaded the images before executing the test and I think that had the same errors.

Defined Variables 
![image](https://github.com/medic/cht-docs/assets/94494491/af283ed8-26f1-4261-a209-e2275e66588e)

First run, no downloading the images -> [upgradeLogs_withoutImagesdownloaded.zip](https://github.com/medic/cht-docs/files/13451916/upgradeLogs_withoutImagesdownloaded.zip)


Second run, with images already downloaded -> [upgradeLogs_withImagesDownloaded.zip](https://github.com/medic/cht-docs/files/13451926/upgradeLogs_withImagesDownloaded.zip)

Downloaded images:
![image](https://github.com/medic/cht-docs/assets/94494491/00f9e03f-a8e1-4f3d-b0cf-a437cd32f2da)

Please let me know if there is anything else I could try.
Thanks @dianabarsan!

Reply:
My guess is that the test doesn't pass the correct path to the docker socket to the upgrade service. This is why I asked if you've ever used the upgrade service separately. 
What OS are you running?

Reply:
macOS Sonoma Version 14.1.1

Reply:
I see that in the script it is defined as `DOCKER_CONFIG_PATH=$homeDir`

Reply:
Can you change that path locally here: https://github.com/medic/cht-core/blob/b8519be2e415deaf20137c80ce55d9257321cff3/tests/e2e/upgrade/wdio.conf.js#L64 and try again?

Reply:
Diana!!! 🤯 

![image](https://github.com/medic/cht-docs/assets/94494491/23d5e598-dca8-44f9-a16f-57a12747e410)

I just changed it to an empty string (because that is the value of `$homeDir`) -> `DOCKER_CONFIG_PATH: '',`

Reply:
Cool! So happy we got to the bottom of this! Thanks for being patient. 
I believe we should just change the code to take this value from an environment variable. I'll try to push a change shortly, and ask you to try the upgrade again (in a few hours) and then we can document the new environment variable.

Reply:
Perfect!
Thank you so much for all the help!! 🌟

Reply:
@tatilepizs @dianabarsan it works!! Thank you:
For me it worked changing to `DOCKER_CONFIG_PATH: os.homedir(),`
I tested manually downloaded the images and without the images.